core/src/main/java/org/infinispan/commands/write/PutMapCommand.java

```diff
@@ -92,13 +92,21 @@ public Object perform(InvocationContext ctx) throws Throwable {
       Map<Object, Object> previousValues = new HashMap<>();
       for (Entry<Object, Object> e : map.entrySet()) {
          Object key = e.getKey();
-         MVCCEntry me = lookupMvccEntry(ctx, key);
-         if (me != null) {
-            Object value = me.getValue();
-            previousValues.put(key, value);
-            notifier.notifyCacheEntryModified(key, e.getValue(), metadata, value, me.getMetadata(), true, ctx, this);
-            me.setValue(e.getValue());
-            me.setChanged(true);
+         Object newValue = e.getValue();
+         MVCCEntry contextEntry = lookupMvccEntry(ctx, key);
+         if (contextEntry != null) {
+            Object previousValue = contextEntry.getValue();
+            previousValues.put(key, previousValue);
+            Metadata previousMetadata = contextEntry.getMetadata();
+            if (contextEntry.isCreated()) {
+               notifier.notifyCacheEntryCreated(key, newValue, metadata, true, ctx, this);
+            }
+            // TODO Only invoke the modified listener if the entry was not created?
+            notifier.notifyCacheEntryModified(key, newValue, metadata, previousValue, previousMetadata, true, ctx,
```

Comment:
You'd think so, but `PutMapCommand` does both, and there's even a test for it: `CacheNotifierTxTest`.
I didn't want to include externally-visible changes in this PR, but I've created https://issues.jboss.org/browse/ISPN-5752 to track this.